### ISSUE TITLE
Add ability to define custom entries in pg_hba.conf

### ISSUE THREAD

**Issue by sgotti** [sorintlab/stolon#110]
Add a new cluster config field for custom entries in pg_hba.conf

**Comment by sajal** [comment]
Any updates on this?
I am trying to experiment with wal-e. The `backup-push` command expects wal-e to be able to logon using local socket which needs `local all all trust` in pg_hba.conf . To experiment with I changed the file by hand in the master and ran `SELECT pg_reload_conf();` which made `backup-push` work as expected. 

One way might be to hardcode the desired config in `pkg/postgresql/postgresql.go` but I'm looking for a more elegant and maintainable solution.

**Comment by sgotti** [comment]
@sajal about this issue it needs more thoughts on how to integrate the stolon required entries with user provided ones.

About wal-e you can just define the connection parameters exporting the libpq environment variables: PGHOST, PGUSER, PGPASSFILE before invoking it.

**Comment by sajal** [comment]
@sgotti Thanks a lot works with those env set correctly.

**Comment by sgotti** [comment]
@sajal Great. I don't remember if the wal-e documentation explains this, if not it'll be probably worth to open a PR there.

**Comment by smkingsoft** [comment]
In pg_hba.conf, the set like
 `host all all 0.0.0.0/0 md5
host all all ::0/0 md5`
is not safe.

can i use Cluster Specification set some custom sets?
replace or append these sets.
eg.
`host all all 10.10.10.0/16 md5`

**Comment by sgotti** [comment]
> can i use Cluster Specification set some custom sets?

Currently not, this issue was opened to track this enhancements.


> In pg_hba.conf, the set like
> host all all 0.0.0.0/0 md5 host all all ::0/0 md5
> is not safe.

I won't say it's not safe. It's just the best trade off we found to work with multiple different environments until we have the ability to set custom rules. Many people don't use postgres host based access but external filtering like host firewall, kubernetes ingress controllers, aws security groups etc... You could also enable ssl communication (but it's probably currently not possible to force it to sslonly).

Some history and thoughts:

Currently the pg_hba.conf is generated by stolon and it looks like this:

```
local all all md5
host all all 0.0.0.0/0 md5
host all all ::0/0 md5
host replication repluser 0.0.0.0/0 md5
host replication repluser ::0/0 md5
```

The first line (local md5) is needed by the commands executed by the keeper on the managed postgres instance (and could be improved and fine tuned).

The last lines are needed for replication between the postgres instances in the cluster

The 2 and 3rd line are for generic access from external clients. In environments like k8s the client IPs (using a replication controller) are dynamic and change at every pod creation so it's not possible to know them ahead of time.

What we could improve:

* Ability to set external client access rules (the 2 and 3rd line). This will make possible to use different authentication mechanism (like client certificates)

* Ability to set the replication rules to only the known keepers ip exposed addresses.

* Use local socket instead of network access for keeper access to local instance.

I'll be happy to hear you thoughts and additional proposal and use cases (and also accept PRs).

**Comment by vbasem** [comment]
@sgotti those recommendations are excellent for environments where certain security policies are enforced. For me it is (almost) required that we use PAM instead of MD5 and limit the IP for access (both for replication and general access).

Additionally we would like to introduce multiple users with different restrictions (specially DB level) which can only originate from certain IPs

Anyone willing to work on this issue?

**Comment by sgotti** [comment]
> limit the IP for access (both for replication and general access).

as an idditional note: limiting IPs for general access is probably not possible since your clients have to connect to the stolon proxies that will forward the data to the right primary (they are simple layer 4 proxies, so they won't talk the postgres wire protocol, inspect data, change data). So the postgres instance will see the proxies IPs.

**Comment by emdem** [comment]
@sgotti Would it be acceptable to make it possible to setup/change the pg_hba.conf with a method similar to the pgParameters?

I have a somewhat rough working implementation that uses a simple string array.

for example:
`stolonctl init '{ "initMode": "new", "pgParameters": {"log_min_duration_statement": "1s" }, "pgHbaConfig": ["local all all md5", "host all all 0.0.0.0/0 md5", "host all all ::0:0 trust", "host replication repluser 0.0.0.0/0 md5", "host replication repluser ::0/0 md5"]}' --cluster-name=cluster --store-backend=consul --store-endpoints=http://consul:8500`

**Comment by sgotti** [comment]
@emdem probably it could be defined in that way. Obviously the first entries must be set by stolon or they will break the replication between the instances.

**Comment by smkingsoft** [comment]
@sgotti the user may provides these hba.conf parameters in the way like @emdem said. 
If the user does not set these parameters, the default parameters are used as stolon set before?

**Comment by sgotti** [comment]
@emdem yes that's the idea. The first entries (for keeper local connections and replication between instances) are manager by stolon so they'll match before the user provided entries. If there aren't user provided entries then the current default (match all dbs and users with md5 auth) will be used.

As an initial step yesterday I opened PR #337 .

**Comment by sgotti** [comment]
Opened PR #341

**Comment by sgotti** [comment]
FIxed in #341
